### ISSUE TITLE
Calling `Branch.edit_protection` fails when any `required_pull_request_reviews` field set

### ISSUE THREAD

**Issue by jdkandersson** [PyGithub/PyGithub#2578]
<strike>It seems that the signature for `Branch.edit_protection` differs in the `.pyi` and `.py` file. In the [`.pyi`](https://github.com/PyGithub/PyGithub/blob/094538e10c1da6f3bbd31ea1f2d102e48e3649c5/github/Branch.pyi#L38) file the arguments are called `user_bypass_pull_request_allowances`, `team_bypass_pull_request_allowances` and `app_bypass_pull_request_allowances` (not the leading singular) whereas in the [.py`](https://github.com/PyGithub/PyGithub/blob/094538e10c1da6f3bbd31ea1f2d102e48e3649c5/github/Branch.py#L131C4-L131C4) file the arguments are plural (`users_bypass_pull_request_allowances`, `teams_bypass_pull_request_allowances` and `apps_bypass_pull_request_allowances`)</strike> fixed in #2571.

There also seems to be a problem where passing nothing for `user_bypass_pull_request_allowances` results in an error when `require_code_owner_reviews` or `dismiss_stale_reviews` are given:

```
github.GithubException.GithubException: 422 {"message": "Invalid request.\n\nNo subschema in \"anyOf\" matched.\nFor 'properties/bypass_pull_request_allowances', nil is not an object.\nNot all subschemas of \"allOf\" matched.\nFor 'anyOf/1', {\"dismiss_stale_reviews\"=>true, \"require_code_owner_reviews\"=>true, \"bypass_pull_request_allowances\"=>nil} is not a null.", "documentation_url": "https://docs.github.com/rest/branches/branch-protection#update-branch-protection"}
```

https://docs.github.com/en/rest/branches/branch-protection?apiVersion=2022-11-28#update-branch-protection

**Comment by EnricoMi** [comment]
You are right, arguments `users_bypass_pull_request_allowances`, `teams_bypass_pull_request_allowances`, `apps_bypass_pull_request_allowances` are wrong in `pyi` and argument `block_creations` is missing. This is going to be fixed in #2571.

**Comment by EnricoMi** [comment]
> There also seems to be a problem where passing nothing for `user_bypass_pull_request_allowances` ...

What do you mean with `nothing`, did you set `user_bypass_pull_request_allowances=None` or did you not have the `user_bypass_pull_request_allowances` argument in the method call at all?

**Comment by jdkandersson** [comment]
> > There also seems to be a problem where passing nothing for `user_bypass_pull_request_allowances` ...
> 
> What do you mean with `nothing`, did you set `user_bypass_pull_request_allowances=None` or did you not have the `user_bypass_pull_request_allowances` argument in the method call at all?

Not having it in the method call at all, e.g.,

```Python
branch.edit_protection(
    require_code_owner_reviews=False,
    dismiss_stale_reviews=False
)
```

Causes the 422 whereas this does work:

```Python
branch.edit_protection(
    require_code_owner_reviews=False,
    dismiss_stale_reviews=False,
    users_bypass_pull_request_allowances=[]
)
```

**Comment by jdkandersson** [comment]
I also noticed that 

```python
branch.edit_protection(
    require_code_owner_reviews=False,
    dismiss_stale_reviews=False,
    users_bypass_pull_request_allowances=[]
)
```

Doesn't work for forks, there is a 422 error which says

```
github.GithubException.GithubException: 422 {"message": "Validation Failed", "errors": ["Only organization repositories can have users and team restrictions"], "documentation_url": "https://docs.github.com/rest/branches/branch-protection#update-branch-protection"}
```

**Comment by EnricoMi** [comment]
Alright, mentioning any of `require_code_owner_reviews` and `dismiss_stale_reviews` adds the `required_pull_request_reviews` field to the requests.

https://github.com/PyGithub/PyGithub/blob/094538e10c1da6f3bbd31ea1f2d102e48e3649c5/github/Branch.py#L245-L255

The `bypass_pull_request_allowances ` field is required but currently omitted when none of users, teams or apps bypass are given.

https://github.com/PyGithub/PyGithub/blob/094538e10c1da6f3bbd31ea1f2d102e48e3649c5/github/Branch.py#L292-L297

Maybe an empty object suffices (as it does for the `dismissal_restrictions` field).

That `dismissal_restrictions` field also mentions
> User and team dismissal_restrictions are only available for organization-owned repositories. Omit this parameter for personal repositories.

There is no such note for `bypass_pull_request_allowances`, but your observation indicates that. To give full control over whether field `users`, `teams` or `apps` exist in the `bypass_pull_request_allowances` object, they should only be added to the `bypass_pull_request_allowances` object when set in the method call.

https://docs.github.com/en/rest/branches/branch-protection?apiVersion=2022-11-28#update-branch-protection

**Comment by jdkandersson** [comment]
I'll see if I can have a go

**Comment by EnricoMi** [comment]
Looks like there is a pull request already: #2535

**Comment by trim21** [comment]
> You are right, arguments `users_bypass_pull_request_allowances`, `teams_bypass_pull_request_allowances`, `apps_bypass_pull_request_allowances` are wrong in `pyi` and argument `block_creations` is missing. This is going to be fixed in #2571.

doesn't look like 2571 will fix this

**Comment by jodelasur** [comment]
Hi! Any chance we can have the fix for this available in 1.59? We recently updated and now encounter this issue.
